### PR TITLE
GH-1905: Implemented the file download functionality.

### LINKED ISSUE #1905: Allow to download files from the workspace
A command that allows to download selected files from the workspace.
It should automatically tar multi-selections (could be done in a second PR).

See original discussion #1874

### PR DESCRIPTION
**Pull request by kittaakos**
Closes #1905.

Signed-off-by: Akos Kitta <kittaakos@gmail.com>

### CODE REVIEW

**Comment by kittaakos** [comment]
It is ready for the review.

File: packages/file-download/src/node/file-download-endpoint.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import { injectable, inject, named } from 'inversify';
+import { json } from 'body-parser';
+import { Application, Router } from 'express';
+import { BackendApplicationContribution } from '@theia/core/lib/node/backend-application';
+import { FileDownloadHandler } from './file-download-handler';
+
+@injectable()
+export class FileDownloadEndpoint implements BackendApplicationContribution {
+
+    protected static PATH = '/file-download';
```
Comment (svenefftinge) [inline packages/file-download/src/node/file-download-endpoint.ts]: Should be just `files`.

File: packages/file-download/src/browser/file-download-command-contribution.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    protected isValid(uris: URI[]): boolean {
+        if (uris.length === 0) {
+            return false;
+        }
+        if (uris.length === 1) {
+            return true;
+        }
+        // Can download multiple files iff they are from the same container folder.
```
Comment (svenefftinge) [inline packages/file-download/src/browser/file-download-command-contribution.ts]: Why restrict it like that?

File: packages/file-download/src/node/file-download-endpoint.ts (hunk 1 above)
Comment (kittaakos) [reply packages/file-download/src/node/file-download-endpoint.ts]: 👍

File: packages/file-download/src/browser/file-download-command-contribution.ts (hunk 2 above)
Comment (kittaakos) [reply packages/file-download/src/browser/file-download-command-contribution.ts]: I am OK removing the restriction if you could help me define a clear way how the files should be zipped up in a folder when selecting multiple files "randomly" from a workspace.

**Comment by svenefftinge** [comment]
I would prefer it. Each extension adds a lot of infrastructural overhead (and degrades build performance) so it needs to be balanced when we create a new extension. I think in this case it tends towards having it in file-system as it is an endpoint exposing it.

File: packages/file-download/src/browser/file-download-command-contribution.ts (hunk 2 above)
Comment (svenefftinge) [reply packages/file-download/src/browser/file-download-command-contribution.ts]: We could preserve the structure and use the common parent of all files as the container.

**Comment by kittaakos** [comment]
> Each extension adds a lot of infrastructural overhead (and degrades build performance)

It is unfortunate, we have to define extensions and adapt the code to this, but I see your point.

**Comment by svenefftinge** [comment]
Also, could we use tar instead of zip?
Zipping takes too long and blocks the node process. Also Users don't get any kind of progress feedback. Using tar instead makes that part fast and moves the work to the download, which has already a nice progress :).

File: packages/file-download/src/browser/file-download-command-contribution.ts (hunk 2 above)
Comment (kittaakos) [reply packages/file-download/src/browser/file-download-command-contribution.ts]: I had the same idea, then I thought; we have to overcomplicate the backend logic for a feature which we won't use that frequently anyway, but let's do.

**Comment by kittaakos** [comment]
> Also, could we use tar instead of zip?

What can I do with a tar as a Windows user? I am just asking... I am not aware of any built-in tools that unpack the tar on Windows. Is it a good idea to force users to use 7Zip?

**Comment by svenefftinge** [comment]
@kittaakos did you see the build failures?

**Comment by kittaakos** [comment]
Thanks for the note: no. I am still in the middle of the changes; the commits are not yet squashed... I will ping you here when it is ready for the review.

**Comment by kittaakos** [comment]
@svenefftinge, I have made a few changes regarding the multi-file download feature; I would consider this PR ready for review.

**Comment by svenefftinge** [comment]
Downloading a directory (e.g. packages in theia repo) is still taking far too long to go without any kind of user feedback. We should at least add an item to the statusbar : `preparing download ...` with some spinning icon or so until the response comes in.

**Comment by kittaakos** [comment]
> We should at least add an item to the statusbar : preparing download ... with some spinning icon or so until the response comes in.

OK.

**Comment by kittaakos** [comment]
I did some measurements, and the bottleneck is the archiving part: this takes 29430 ms when downloading the `node_modules` folder (573.2 MB) from Theia. I will try to pipe the archiving stream into the response. Hopefully, it makes the user experience smoother, if no, I update the status bar to give some user feedback.

**Comment by kittaakos** [comment]
I have managed to pipe the `tar` stream into the response but the download progress indicator did not show up in the browser even with [`downloadjs`](https://www.npmjs.com/package/downloadjs). I go with the status bar approach.

**Comment by kittaakos** [comment]
> We should at least add an item to the statusbar : preparing download ... with some spinning icon or so until the response comes in.

This is done.

**Comment by kittaakos** [comment]
I tried to fine-tune the status bar's `Preparing download...` visibility. It is still not perfect; see the attached screen-casts below. I have already moved the status bar updating logic to the very last possible position where I still do "have control" over the download.

This could be the root of the defect (source: https://github.com/whatwg/html/issues/954#issue-144165132):
<i>
[...]The problem is that browsers tend to start the download asynchronously after the link has been clicked. This is especially browsers which show the user a "do you want to open or download this file?" dialog, where the download might start long after the link was clicked.[...]
</i>

Downloading `packages`:
gh-1905-packages-download

Downloading `node_modules` (cut):
gh-1905-node_modules-download-cut

Downloading `theia` (cut):
gh-1905-theia-download-cut

**Comment by davide-sergi** [comment]
Is there a flow diagram somewhere in order to understand how download a file/folder programmatically?

**Comment by kittaakos** [comment]
> flow diagram

No, there isn't. At least, I am not aware of it.

> how download a file/folder programmatically?

Check [this](https://github.com/eclipse-theia/theia/issues/5707#issuecomment-511390894) thread. It might be outdated, though. If it does not work now, please open a follow-up issue, or start a discussion, to attract more attention for your problem. This is a closed PR.

I hope this helps.

**Pull request merged**